packages/nimble-components/src/table/index.ts

```diff
@@ -161,6 +212,60 @@ export class Table<
 
         this.updateTableOptions({ columns: generatedColumns });
     }
+
+    private updateVirtualizer(): void {
+        const options = this.createVirtualizerOptions();
+        if (this.virtualizer) {
+            this.virtualizer.setOptions(options);
+        } else {
+            this.virtualizer = new Virtualizer(options);
+        }
+        this.virtualizer._willUpdate();
+        this.handleVirtualizerChange();
+    }
+
+    private createVirtualizerOptions(): VirtualizerOptions<
+    HTMLElement,
+    HTMLElement
+    > {
+        const rowHeight = parseFloat(controlHeight.getValueFor(this));
+        return {
+            count: this.tableData.length,
+            getScrollElement: () => {
+                return this.viewport;
+            },
+            estimateSize: (_: number) => rowHeight,
+            enableSmoothScroll: true,
+            overscan: 3,
+            scrollToFn: elementScroll,
+            observeElementOffset,
+            observeElementRect,
+            onChange: () => this.handleVirtualizerChange()
+        } as VirtualizerOptions<HTMLElement, HTMLElement>;
+    }
+
+    private handleVirtualizerChange(): void {
+        const virtualizer = this.virtualizer!;
+        this.visibleItems = virtualizer.getVirtualItems();
+        this.allRowsHeight = virtualizer.getTotalSize();
+        // We're using a separate div ('table-scroll') to represent the full height of all rows, and
+        // the row container's height is only big enough to hold the virtualized rows. So we don't
+        // use the TanStackVirtual-provided 'start' offset (which is in terms of the full height)
+        // to translate every individual row, we just translate the row container.
+        let rowContainerYOffset = 0;
+        if (this.visibleItems.length > 0) {
+            const firstItem = this.visibleItems[0]!;
+            const lastItem = this.visibleItems[this.visibleItems.length - 1]!;
+            if (lastItem.end < this.allRowsHeight) {
+                rowContainerYOffset = firstItem.start - this.viewport.scrollTop;
+            }
+        }
+        this.rowContainerYOffset = rowContainerYOffset;
+        // If we have enough rows that a vertical scrollbar is shown, we need to offset the header widths
+        // by the same margin so the column headers align with the corresponding rendered cells
+        this.headerContainerMarginRight = this.viewport.getBoundingClientRect().width
```

Comment:
onChange is going to be called rapidly right? Like continuously during scroll / on the order of raF?

I'm concerned about getBoundingClientRect because will trigger a layout. Although after looking at what triggers layout so may reading [scrolltop and scrollwidth](https://gist.github.com/paulirish/5d52fb081b3570c81e3a).

I didn't look closely in devtools yet but can we make sure we aren't layout thrashing (updating dom then reading something that triggers layout then updating dom, etc).

Also can we potentially cache off some of the dimensions we need with resize observer? I assume we can with the viewport sizing used here, but that might not be useful at preventing layout if we can't also cache off scroll state. Does Virtualizer do any ResizeOberservers to trigger onChange? I would imagine so... does it not already have the measurement for this.viewport saved off?

Reply:
Yes, `onChange` is called repeatedly as you scroll.  
The Virtualizer uses a ResizeObserver internally, but it doesn't have public properties/methods for all the measurements in this method. After looking at it's API again I can replace `this.viewport.scrollTop` with `virtualizer.scrollOffset`, but the `getBoundingClientRect().width` and `scrollWidth` code will still be in this method.

Performance seems fine though, and I don't see any signs of layout thrashing after profiling in DevTools:
![TableScroll1](https://user-images.githubusercontent.com/20709258/214429242-f3a722fe-aa05-4261-b538-d20cdafbc8b8.PNG)
![TableScroll2](https://user-images.githubusercontent.com/20709258/214429259-9e2b69d5-b910-4039-9374-a7097e3e3315.PNG)  
Each 'task' is a chunk of the scrolling, and everything is handled in 5ms. onScroll and rAF => animation frame => 1 Recalculate Style => 1 layout => 1 paint which makes sense, and no messages about Forced Reflow or anything like that.

Reply:
Made the 1 minor update that I mentioned. Marking resolved as I don't think that the performance capture shows any concerns.

Reply:
Do we have any guarantee that handleVirtualizerChange will only be called in a ResizeObserver (which runs after layout so it's known to be safe to read values from layout)?

I think it could be called in more cases than that scanning the virtualizer source but it's hard to tell just from source inspection. If we feel confident this won't be called unnecessarily outside of resize then we should maybe comment on the expectations of when this runs.

Alternatively if we are not confident then I think we should have a ResizeObserver on `this.viewport` where we can cache off current viewport bounding rect (already part of the callback params) and get the value of viewport.scrollWidth in the the same resizeobserver callback to cache off as well. We can then used the cached values here.

Reply:
Since I don't think we have guarantees that onChange will only be called from a ResizeObserver, I added our own ResizeObserver for `this.viewport` in the update I just pushed.  
Since needing to recompute the header right margin is closer to a side effect of the scroll/ virtualizer changes, rather than caching off the viewport dimensions, I opted to just compute and set `headerContainerRightMargin` directly in the resize observer callback, which seemed a little simpler/cleaner.